src/inference/Inference.jl

```diff
@@ -77,24 +84,368 @@ function mh_accept(H::T, H_new::T, log_proposal_ratio::T) where {T<:Real}
 end
 
-# Concrete algorithm implementations.
-include("sghmc.jl")
+######################
+# Default Transition #
+######################
+
+struct Transition{T, F<:AbstractFloat} <: AbstractTransition
+    θ  :: T
+    lp :: F
+end
+
+function transition(spl::Sampler, nt::NamedTuple=NamedTuple())
+    theta = merge(tonamedtuple(spl.state.vi), nt)
+    lp = getlogp(spl.state.vi)
+    return Transition{typeof(theta), typeof(lp)}(theta, lp)
+end
+
+function additional_parameters(::Type{Transition})
+    return [:lp]
+end
+
+Interface.transition_type(::Sampler{alg}) where alg = transition_type(alg)
+
+##########################################
+# Internal variable names for MCMCChains #
+##########################################
+
+const TURING_INTERNAL_VARS =
+    (internals = ["elapsed", "eval_num", "lf_eps", "lp", "weight", "le",
+                  "acceptance_rate", "hamiltonian_energy", "is_accept", 
+                  "log_density", "n_steps", "numerical_error", "step_size",
+                  "tree_depth"],)
+
+#########################################
+# Default definitions for the interface #
+#########################################
+
+function Interface.sample(
+    rng::AbstractRNG,
+    model::ModelType,
+    alg::AlgType,
+    N::Integer;
+    kwargs...
+) where {
+    ModelType<:Sampleable,
+    SamplerType<:AbstractSampler,
+    AlgType<:InferenceAlgorithm
+}
+    return sample(rng, model, Sampler(alg, model), N; progress=PROGRESS[], kwargs...)
+end
+
+function Interface.sample(
+    model::ModelType,
+    alg::AlgType,
+    N::Integer;
+    resume_from=nothing,
+    kwargs...
+) where {
+    ModelType<:Sampleable,
+    SamplerType<:AbstractSampler,
+    AlgType<:InferenceAlgorithm
+}
+    if resume_from === nothing
+        return sample(model, Sampler(alg, model), N; progress=PROGRESS[], kwargs...)
+    else
+        return resume(resume_from, N)
+    end
+end
+
+function Interface.sample_init!(
+    ::AbstractRNG,
+    model::Model,
+    spl::Sampler,
+    N::Integer;
+    kwargs...
+)
+    # Resume the sampler.
+    set_resume!(spl; kwargs...)
+
+    # Set the parameters to a starting value.
+    initialize_parameters!(spl; kwargs...)
+end
+
+function Interface.sample_end!(
+    ::AbstractRNG,
+    ::Model,
+    spl::AbstractSampler,
+    ::Integer,
+    ::Vector{TransitionType};
+    kwargs...
+) where {TransitionType<:AbstractTransition}
+    # Silence the default API function.
+end
+
+function initialize_parameters!(
+    spl::AbstractSampler;
+    init_theta::Union{Nothing,Array{<:Any,1}}=nothing,
+    verbose::Bool=false,
+    kwargs...
+)
+    # Get `init_theta`
+    if init_theta != nothing
+        verbose && @info "Using passed-in initial variable values" init_theta
+        # Convert individual numbers to length 1 vector; `ismissing(v)` is needed as `size(missing)` is undefined`
+        init_theta = [ismissing(v) || size(v) == () ? [v] : v for v in init_theta]
```

Comment:
Is this doing the right thing when `ismissing(v)` or `size(v) == ()`? Is it tested?

Reply:
What does it mean for `size(v)` to be `()`? Should we error instead?

Reply:
This is a very tough design by me.

Users can pass initial parameters either by flatten vector or nested vector.

Users should also be allowed to `missing` to indicate some parameters are not provided with initial value. 

I'm open to improve this code by maybe in another PR.

Reply:
Yeah, we should eventually give users the ability to call something like

```julia
sample(model, alg; init_theta=(m=0.2, s=0.4))
```

For right now we'll live with it, so I'll resolve this.